### PR TITLE
feat: Add GitHub GraphQLOptimizer

### LINKED ISSUE #2836: Support GraphQL filter parameters in `pull_requests`

### LINKED ISSUE #2837: Support GraphQL filter parameters in `issues`

### PR DESCRIPTION
**Pull request by peasee**
## 🗣 Description

* Updates the GitHub pulls column `login` to be `author`, to match the expected filter value (and to match the column in issues)
* Adds a `GraphQLOptimizer` for issues and pulls
* Updates the signature of `GraphQLOptimizer::parameter_injection` to pass the entire `GraphQLQuery` instead of just the AST `Document<'_, _>`. This allows the parameter injection to also update the pagination parameters after it has modified the field parameters.

Filter push down in practice results in some good performance gains, depending on the query:

```bash
sql> select author, state, title from spiceai.pulls where state = 'OPEN'
+-----------+-------+----------------------------------------------------+
| author    | state | title                                              |
+-----------+-------+----------------------------------------------------+
| Jeadie    | OPEN  | Update quickstarts for Endgame                     |
| Jeadie    | OPEN  | Add `log/slog` to spice CLI tool                   |
| slyons    | OPEN  | Adding integration test for DuckDB read functions  |
| lukekim   | OPEN  | Add NSQL CLI REPL command                          |
| Sevenannn | OPEN  | Add busy_timeout accelerator param for Sqlite      |
| Sevenannn | OPEN  | Run tpcds for accelerator benchmarks               |
| peasee    | OPEN  | Feat/add filter pushdown support for issues        |
| Jeadie    | OPEN  | Enable `ArrayDistance` to handle arrays of vectors |
+-----------+-------+----------------------------------------------------+

Time: 0.895188481 seconds. 8 rows.
sql> select author, state, title from spiceai.pulls where state = 'OPEN'
+-----------+-------+----------------------------------------------------+
| author    | state | title                                              |
+-----------+-------+----------------------------------------------------+
| Jeadie    | OPEN  | Enable `ArrayDistance` to handle arrays of vectors |
| Sevenannn | OPEN  | Add busy_timeout accelerator param for Sqlite      |
| lukekim   | OPEN  | Add NSQL CLI REPL command                          |
| slyons    | OPEN  | Adding integration test for DuckDB read functions  |
| Jeadie    | OPEN  | Add `log/slog` to spice CLI tool                   |
| Jeadie    | OPEN  | Update quickstarts for Endgame                     |
| peasee    | OPEN  | Feat/add filter pushdown support for issues        |
| Sevenannn | OPEN  | Run tpcds for accelerator benchmarks               |
+-----------+-------+----------------------------------------------------+

Time: 65.83647304 seconds. 8 rows.
```

Other queries are not as drastic, but still good improvements.

## 🔨 Related Issues

* Closes #2837
* Closes #2836

### CODE REVIEW

**Review by sgrebnov — DISMISSED** [review]

File: crates/runtime/src/dataconnector/github/mod.rs [hunk 1]
```diff
@@ -370,3 +396,151 @@ pub fn parse_globs(input: &str) -> super::DataConnectorResult<Arc<GlobSet>> {
...
+    }
+
+    FilterPushdownResult {
+        filter_pushdown: TableProviderFilterPushDown::Unsupported,
+        expr: expr.clone(),
+        context: None,
+    }
+}
+
+pub(crate) fn parameter_injection(
```
Comment (sgrebnov) [inline crates/runtime/src/dataconnector/github/mod.rs]: Nit: verb like `inject_parameters` is better IMO

File: crates/runtime/src/dataconnector/github/mod.rs [hunk 2]
```diff
@@ -370,3 +396,151 @@ pub fn parse_globs(input: &str) -> super::DataConnectorResult<Arc<GlobSet>> {
         .context(super::InvalidGlobPatternSnafu { pattern: input })?;
     Ok(Arc::new(glob_set))
 }
+
+// TODO: add support for LIKE and IN filters, to support columns like assignees, labels, etc
+// TODO: add support for date comparisons to support columns like updated_at, created_at, closed_at, etc
```
Comment (sgrebnov) [inline crates/runtime/src/dataconnector/github/mod.rs:401]: Support for `updated_at` is critical for append refresh. I see this as a common and recommended by us pattern so you fetch updated items only when accelerating table.

Comment (peasee) [reply crates/runtime/src/dataconnector/github/mod.rs:401]: Those columns still work, they just don't have filter push down enabled for now.

Comment (peasee) [reply crates/runtime/src/dataconnector/github/mod.rs:401]: I've raised a follow up issue to ensure this gets tracked: https://github.com/spiceai/spiceai/issues/2893

**Review by sgrebnov — APPROVED** [review]

**Pull request merged**